File: August-LeetCoding-Challenge/13-Iterator-for-Combination/Iterator-for-Combination.py
class CombinationIterator:

    def __init__(self, characters: str, combinationLength: int):
        self.res = []
        if combinationLength==1:
            self.res = [x for x in characters]
        else:
            self.bk([],characters,combinationLength)
        
        
    def bk(self,cur,choice,fin):
        if len(cur) == fin:
            self.res.append("".join(cur))
            return
        for i in range(len(choice)):
            cur.append(choice[i])
            self.bk(cur,choice[i+1:],fin)
            cur.pop(-1)
        return 
        

    def next(self) -> str:
        tmp = self.res[0]
        self.res.pop(0)
        return tmp

    def hasNext(self) -> bool:
        return self.res!=[]

File: August-LeetCoding-Challenge/13-Iterator-for-Combination/Iterator-for-Combination.cs
public class CombinationIterator {

    private Queue<string> queue = new Queue<string>();
    
    public CombinationIterator(string characters, int combinationLength) {
        Dfs(characters, 0, "", combinationLength);
    }
        
    public string Next() {
        return queue.Dequeue();
    }
    
    public bool HasNext() {
        return queue.Count > 0;
    }
    
    private void Dfs(string chars, int start, string curr, int length){
        if(length == 0){
            queue.Enqueue(curr);
        }
        for(int i=start; i<chars.Length; i++){
            Dfs(chars, i+1, curr + chars[i], length-1);
        }
    }
}

/**
 * Your CombinationIterator object will be instantiated and called as such:
 * CombinationIterator obj = new CombinationIterator(characters, combinationLength);
 * string param_1 = obj.Next();
 * bool param_2 = obj.HasNext();
 */